### DISCUSSION TITLE
在此示例中，VLESS+H2C+TLS 不起作用

### DISCUSSION THREAD

**Discussion by tegant** [lxhao61/integrated-examples#220] [Q&A] [answered]
我刚刚测试了  https://github.com/lxhao61/integrated-examples/tree/main/Xray(M%2BF%2BH%2BD%2BG%2BB%2BA)%2BCaddy(N) ，它们都有效，但 VLESS+H2C+TLS 无效。其他都很好。

**Comment by lxhao61** [comment]
示例无问题，请仔细检查自己配置！
另你使用什么客户端？

**Reply by tegant** [reply]
我用的是 [libXray](https://github.com/XTLS/libXray) 编译的 iOS 客户端，在 iPhone 上运行。从 Caddy 的日志来看，出现了 404 错误：

```
status: 404
method: "PUT"
uri: "/475ea934"
host: "www.example.com"  # 这个似乎不对，应该显示我实际配置的域名
```

配置看起来是对的：
1. Xray 配置中设置了正确的 path: "/475ea934"
2. Caddy 配置中也匹配了相同的 path，并正确配置了 h2c 反向代理

但有两个细节可能值得注意：
1. 日志显示客户端在使用 PUT 方法发送请求，而不是常见的 POST 或 GET
2. host 显示为 www.example.com ，这看起来不太对

你们觉得这些异常会造成问题吗？或者建议我再检查哪些地方？

/var/log/caddy/access.log:
> 2024/10/29 14:35:12.952    INFO    http.log.access.log0    handled request    {"request": {"remote_ip": "152.37.80.166", "remote_port": "1595", "client_ip": "152.37.80.166", "proto": "HTTP/2.0", "method": "PUT", "host": "www.example.com", "uri": "/475ea934", "headers": {"Accept-Encoding": ["identity"], "User-Agent": ["Go-http-client/2.0"]}, "tls": {"resumed": false, "version": 771, "cipher_suite": 52393, "proto": "h2", "server_name": "57f5091d.myDomain.uk"}}, "bytes_read": 0, "user_id": "", "duration": 0.000256819, "size": 0, "status": 404, "resp_headers": {"Strict-Transport-Security": ["max-age=31536000; includeSubDomains; preload"], "Server": ["Caddy"], "Alt-Svc": ["h3=\":443\"; ma=2592000"]}}


Xray 配置:
```
        {
            "listen": "127.0.0.1",
            "port": 2005,
            "protocol": "vless",
            "settings": {
                "clients": [
                    {
                        "id": "73dcceae-19b8-4fd9-9e9c-5485a7ea13e2",
                        "email": "73dcceae-19b8-4fd9-9e9c-5485a7ea13e2@VLESS_H2_TLS.io"
                    },
                ],
                "decryption": "none"
            },
            "streamSettings": {
                "network": "h2",
                "httpSettings": {
                    "path": "/475ea934"
                }
            },
            "sniffing": {
                "enabled": true,
                "destOverride": [
                    "http",
                    "tls",
                    "quic"
                ]
            }
        },
```

Caddy 配置:
```
        "srvh2c": {
          "listen": ["127.0.0.1:88"],
          "listener_wrappers": [{
            "wrapper": "proxy_protocol",
            "allow": ["127.0.0.1/32"]
          }],
          "routes": [{
            "match": [{
              "path": ["/475ea934"]
            }],
            "handle": [{
              "handler": "reverse_proxy",
              "transport": {
                "protocol": "http",
                "versions": ["h2c","2"]
              },
              "upstreams": [{
                "dial": "127.0.0.1:2005"
              }]
            }]
          },
          {
            "handle": [{
              "handler": "headers",
              "response": {
                "set": {
                  "Strict-Transport-Security": ["max-age=31536000; includeSubDomains; preload"]
                }
              }
            },
            {
              "handler": "file_server",
              "root": "/var/www/html"
            }]
          }],
          "logs": {
            "default_logger_name": "log0"
          },
          "protocols": ["h1","h2c"]
        },
```

**Reply by lxhao61** [reply, marked as answer]
所述的细节是没有问题的，默认是这样的。
应该是 IOS 客户端问题，类似 https://github.com/lxhao61/integrated-examples/discussions/107 问题。
服务端与客户端都加 host 配置试试。

**Reply by tegant** [reply]
啊谢谢。看到这个 issue #107，我明白了是 host 的问题。

我试了两种配置：

1. 加上实际域名:
```
"streamSettings": {
    "network": "h2",
    "httpSettings": {
        "host": ["7dedd9dc.myDomain.uk"],
        "path": "/475ea934"
    }
}
```
2. 按照讨论里的建议用空字符串:
```
"streamSettings": {
    "network": "h2",
    "httpSettings": {
        "host": [""],
        "path": "/475ea934"
    }
}
```
但两种方式都还是不行。我是不是还漏了什么其他配置？从 issue 看这应该是客户端和服务端的 host 配置需要匹配，但我这边改了客户端的配置后依然报 404。

**Reply by lxhao61** [reply]
host 配置的域名是 Trojan+RAW+TLS 配置证书的域名吗？

**Reply by tegant** [reply]
谢谢你的帮助。终于解决了。我遇到了两个问题：

1) 第一个问题确实如 [#107](https://github.com/lxhao61/integrated-examples/discussions/107) 一样，服务端与客户端都需要添加 host 配置。也许可以考虑在这个[位置](https://github.com/lxhao61/integrated-examples/blob/09d93167bd2209e8a24a4a8be0d9eb94308b380b/Xray(M%2BF%2BB%2BD%2BG%2BA)%2BCaddy(N)/1_xray_config.json#L135)的示例代码中加入这个说明。

2) 另一个问题是域名。我记得几个月前 VLESS+H2C+TLS 使用的是[这里](https://github.com/lxhao61/integrated-examples/blob/09d93167bd2209e8a24a4a8be0d9eb94308b380b/Xray(M%2BF%2BB%2BD%2BG%2BA)%2BCaddy(N)/1_caddy.json#L224)定义的非 AES 链接，但现在似乎改成使用 Trojan 的链接了。我之前不知道这个变化。

现在一切都正常工作了。非常感谢你的帮助。

**Reply by lxhao61** [reply]
客气了。
1、目前仅 IOS 客户端出现这个问题，其它客户端一切正常。
2、调整由 Trojan+RAW+TLS 负责 TLS，主要是考虑各应用均衡。
